src/overmap.cpp

```diff
@@ -1726,6 +1769,45 @@ void overmap::generate(const overmap *north, const overmap *east,
         }
     }
 
+    if (railroads_out.size() < 2) {
+        std::vector<city> viable_railroads;
+        int tmp;
+        // Populate viable_railroads with one point for each neighborless side.
+        // Make sure these points don't conflict with rivers.
+        // TODO: In theory this is a potential infinte loop...
+        if (north == NULL) {
+            do {
+                tmp = rng(10, OMAPX - 11);
+            } while (is_river(ter(tmp, 0, 0)) || is_river(ter(tmp - 1, 0, 0)) ||
+                     is_river(ter(tmp + 1, 0, 0)) );
+            viable_railroads.push_back(city(tmp, 0, 0));
+        }
+        if (east == NULL) {
+            do {
+                tmp = rng(10, OMAPY - 11);
+            } while (is_river(ter(OMAPX - 1, tmp, 0)) || is_river(ter(OMAPX - 1, tmp - 1, 0)) ||
+                     is_river(ter(OMAPX - 1, tmp + 1, 0)));
+            viable_railroads.push_back(city(OMAPX - 1, tmp, 0));
+        }
+        if (south == NULL) {
+            do {
+                tmp = rng(10, OMAPX - 11);
+            } while (is_river(ter(tmp, OMAPY - 1, 0)) || is_river(ter(tmp - 1, OMAPY - 1, 0)) ||
+                     is_river(ter(tmp + 1, OMAPY - 1, 0)));
+            viable_railroads.push_back(city(tmp, OMAPY - 1, 0));
+        }
+        if (west == NULL) {
+            do {
+                tmp = rng(10, OMAPY - 11);
+            } while (is_river(ter(0, tmp, 0)) || is_river(ter(0, tmp - 1, 0)) ||
+                     is_river(ter(0, tmp + 1, 0)));
+            viable_railroads.push_back(city(0, tmp, 0));
+        }
```

Comment:
4 directions, all handled with hardcoded copypasta.
You should handle them all in a neat loop.